utils/tinystr/src/unvalidated.rs

```diff
@@ -0,0 +1,88 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::TinyAsciiStr;
+use crate::TinyStrError;
+
+#[derive(Debug, PartialEq, PartialOrd, Eq, Ord, Clone, Copy)]
+pub struct UnvalidatedTinyAsciiStr<const N: usize>(pub(crate) [u8; N]);
+
+impl<const N: usize> UnvalidatedTinyAsciiStr<N> {
+    pub fn try_into_tinystr(&self) -> Result<TinyAsciiStr<N>, TinyStrError> {
+        TinyAsciiStr::try_from_raw(self.0)
+    }
+}
+
+impl<const N: usize> TinyAsciiStr<N> {
+    pub fn raw(self) -> UnvalidatedTinyAsciiStr<N> {
+        UnvalidatedTinyAsciiStr(*self.all_bytes())
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<const N: usize> serde::Serialize for UnvalidatedTinyAsciiStr<N> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::Error;
+        self.try_into_tinystr()
+            .map_err(|_| S::Error::custom("invalid ascii in UnvalidatedTinyAsciiStr"))?
+            .serialize(serializer)
+    }
+}
+
+macro_rules! deserialize {
+    ($size:literal) => {
+        #[cfg(feature = "serde")]
+        impl<'de, 'a> serde::Deserialize<'de> for UnvalidatedTinyAsciiStr<$size>
+        where
+            'de: 'a,
+        {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                if deserializer.is_human_readable() {
+                    Ok(TinyAsciiStr::deserialize(deserializer)?.raw())
+                } else {
+                    Ok(Self(<[u8; $size]>::deserialize(deserializer)?))
```

Comment:
Nit: Please manually invoke `deserialize_bytes` because otherwise this might use `deserialize_seq` which is probably not as efficient.

Reply:
I believe because they're fixed-size, arrays go through `deserialize_tuple`